vmm/src/serial_manager.rs

```diff
@@ -0,0 +1,235 @@
+    Stdin(#[source] vmm_sys_util::errno::Error),
+
+    /// Cannot handle the VM pty stream
+    #[error("Error handling VM pty: {0:?}")]
+    Pty(#[source] io::Error),
+
+    /// Cannot queue input to the serial device
+    #[error("Error queuing input to the serial device: {0}")]
+    QueueInput(#[source] vmm_sys_util::errno::Error),
+}
+pub type Result<T> = result::Result<T, Error>;
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+#[repr(u64)]
+pub enum EpollDispatch {
+    Stdin = 0,
+    SerialPty = 1,
+    Unknown,
+}
+
+impl From<u64> for EpollDispatch {
+    fn from(v: u64) -> Self {
+        use EpollDispatch::*;
+        match v {
+            0 => Stdin,
+            1 => SerialPty,
+            _ => Unknown,
+        }
+    }
+}
+
+pub struct EpollContext {
+    epoll_file: File,
+}
+
+impl EpollContext {
+    pub fn new() -> result::Result<EpollContext, io::Error> {
+        let epoll_fd = epoll::create(true)?;
+        // Use 'File' to enforce closing on 'epoll_fd'
+        let epoll_file = unsafe { File::from_raw_fd(epoll_fd) };
+
+        Ok(EpollContext { epoll_file })
+    }
+
+    pub fn add_stdin(&mut self) -> result::Result<(), io::Error> {
+        let dispatch_index = EpollDispatch::Stdin as u64;
+        epoll::ctl(
+            self.epoll_file.as_raw_fd(),
+            epoll::ControlOptions::EPOLL_CTL_ADD,
+            libc::STDIN_FILENO,
+            epoll::Event::new(epoll::Events::EPOLLIN, dispatch_index),
+        )?;
+
+        Ok(())
+    }
+
+    fn add_event<T>(&mut self, fd: &T, token: EpollDispatch) -> result::Result<(), io::Error>
+    where
+        T: AsRawFd,
+    {
+        let dispatch_index = token as u64;
+        epoll::ctl(
+            self.epoll_file.as_raw_fd(),
+            epoll::ControlOptions::EPOLL_CTL_ADD,
+            fd.as_raw_fd(),
+            epoll::Event::new(epoll::Events::EPOLLIN, dispatch_index),
+        )?;
+
+        Ok(())
+    }
+}
+
+impl AsRawFd for EpollContext {
+    fn as_raw_fd(&self) -> RawFd {
+        self.epoll_file.as_raw_fd()
+    }
+}
+
+pub struct SerialManager {
+    #[cfg(target_arch = "x86_64")]
+    serial: Arc<Mutex<Serial>>,
+    #[cfg(target_arch = "aarch64")]
+    serial: Arc<Mutex<Pl011>>,
+    epoll: EpollContext,
+    pty: Option<File>,
+}
+
+impl SerialManager {
+    #[cfg(target_arch = "x86_64")]
+    pub fn new(serial: Arc<Mutex<Serial>>, pty: Option<File>, serial_tty: bool) -> Result<Self> {
+        let mut epoll = EpollContext::new().map_err(Error::Epoll)?;
+        let pty = if pty.is_some() {
+            let pty_clone = pty.unwrap().try_clone().map_err(Error::FileClone)?;
+            epoll
+                .add_event(&pty_clone, EpollDispatch::SerialPty)
+                .map_err(Error::Epoll)?;
+            Some(pty_clone)
+        } else {
+            None
+        };
+        if serial_tty && unsafe { libc::isatty(libc::STDIN_FILENO as i32) } != 0 {
+            epoll.add_stdin().map_err(Error::Epoll)?;
+        }
```

Comment:
You can't have the serial attached to PTY and TTY so you should a match on the mode from the `ConsoleConfig`